### PR TITLE
Tensorflow workshop updates 

### PR DESCRIPTION
**Pull request by sofia-geislinger**
Read, reviewed, edited workshops. 

Note: The hyperlink to the answer key in the 'wrap-up' page does not work. It didn't work previously, and I tried to see how other workshops implemented the answer key page but didn't get much insight. I can keep trying, but for now the hackathon workshop review is completed.

### CODE REVIEW

File: content/english/python-tensorflow/activity-7.md [hunk 1]
```diff
@@ -9,68 +9,68 @@ draft: false
...
-# Grab an image from the test dataset. This shows the resolution of the image. 
+![Plotting Model Figure 2](../media/Plotting_fig2.png)
 
-#NOTE: This index will be changed and its corresponding plot would be displayed in the next few steps
-img = test_images[0]
+</br>
+We will adjust the labeling of the x-axis so that the data it displays is easier to understand. 
+</br>
+</br>
+Copy and Paste the following code into Google Colab:
```
Comment (ozhang22) [inline content/english/python-tensorflow/activity-7.md:23]:
```suggestion
Copy and paste the following code into your Google Colab notebook:
```

File: content/english/python-tensorflow/activity-7.md [hunk 2]
```diff
@@ -9,68 +9,68 @@ draft: false
...
 
-print(img.shape)
+```python
+img = test_images[0] # Grab an image from the test dataset. 
+#NOTE: This index will be changed and its corresponding plot will be displayed in the next few steps
```
Comment (ozhang22) [inline content/english/python-tensorflow/activity-7.md:27]:
```suggestion
# NOTE: This index will be changed and its corresponding plot will be displayed in the next few steps
```

File: content/english/python-tensorflow/activity-7.md [hunk 3]
```diff
@@ -9,68 +9,68 @@ draft: false
...
+img = (np.expand_dims(img,0)) # This expands the img array 
 
 print(img.shape)
 ```
 
 ```python
-predictions_single = probability_model.predict(img)
+predictions_single = probability_model.predict(img) # This gives the 
+# confidence level that the image corresponds with each category. 
+# For examaple the possibility that the image is a T-shirt is 5.2198538e-07
```
Comment (ozhang22) [inline content/english/python-tensorflow/activity-7.md:42]:
```suggestion
# This gives the confidence level that the image corresponds with each category. 
# For example, the probability that the image is a T-shirt is 5.2198538e-07.
predictions_single = probability_model.predict(img)
```

File: content/english/python-tensorflow/activity-7.md [hunk 4]
```diff
@@ -9,68 +9,68 @@ draft: false
...
 _ = plt.xticks(range(10), class_names, rotation=45)
 ```
 
-{{% notice note %}}
-To verify the index value with the highest probability, we use
+To verify the index value with the highest probability, we use the following code. 
 
 ```python
-np.argmax(predictions_single[0]) #Verifying the index value with highest probability
+np.argmax(predictions_single[0]) # The output of this tells us what category the clothing item had the highest probability of being a part of
```
Comment (ozhang22) [inline content/english/python-tensorflow/activity-7.md:58]:
```suggestion
# The output of this tells us what category the clothing item had the 
# highest probability of being a part of
np.argmax(predictions_single[0]) 
```

File: content/english/python-tensorflow/answer_key.md [hunk 5]
```diff
@@ -1,26 +1,33 @@
 ---
 title: "Answer Key"
 date: 2020-09-14T16:19:17-07:00
-weight: 14
+weight: 15
 draft: false
-hidden: true
+#hidden: true
```
Comment (ozhang22) [inline content/english/python-tensorflow/answer_key.md:6]: Please revert this change - the answer key is intended to be hidden
```suggestion
hidden: true
```

File: content/english/python-tensorflow/building_neural_network.md [hunk 6]
```diff
@@ -9,27 +9,29 @@ draft: false
...
 
-This is the first step in building a neural network, which is actually a tool to allow us to do "machine learning". With this tool, we are training a model/computer to analyze various training images/data, likely labeled in advance, and perform and execute a specific task as a result. In our scenario, we are using pre-labeled images of clothing to train our model to correctly predict the type of clothing for any clothing sample we input.
+This is the first step in building a neural network, which is actually a tool that allows us to do "machine learning". With this tool, we are training a program to analyze images or data, otherwise known as training data. 
+Then after training, we can ask the program to perform and execute a specific task using what it learned. For example, say we train a program with dog and cat pictures and we provide the program with the correct identification of the images. Once trained, we can ask the computer to identify if a picture of my pet Whiskers is of a cat or dog.
```
Comment (ozhang22) [inline content/english/python-tensorflow/building_neural_network.md:13]:
```suggestion
Then after training, we can ask the program to perform and execute a specific task using what it learned. For example, let's say we train a program with dog and cat pictures, and we provide the program with the correct identification of the images. Once trained, we can ask the computer to identify if a picture of my pet Whiskers is of a cat or dog. 
```

File: content/english/python-tensorflow/building_neural_network.md [hunk 7]
```diff
@@ -9,27 +9,29 @@ draft: false
...
+This is the first step in building a neural network, which is actually a tool that allows us to do "machine learning". With this tool, we are training a program to analyze images or data, otherwise known as training data. 
+Then after training, we can ask the program to perform and execute a specific task using what it learned. For example, say we train a program with dog and cat pictures and we provide the program with the correct identification of the images. Once trained, we can ask the computer to identify if a picture of my pet Whiskers is of a cat or dog. 
 
-{{% notice info %}}
-### Neural Network
+ In our scenario, we are using pre-labeled images of clothing to train our model. Then, we want the program to correctly predict the type of clothing an input may be.
```
Comment (ozhang22) [inline content/english/python-tensorflow/building_neural_network.md:15]:
```suggestion
In our scenario, we are using pre-labeled images of clothing to train our model. Then, we want the program to correctly predict the type of clothing an input may be.
```

File: content/english/python-tensorflow/building_neural_network.md [hunk 8]
```diff
@@ -53,15 +55,15 @@ We would recommend looking at the first code block below and the neural network
 - The neurons (128 of them) then recognize different patterns in the images by observing different features and the relationships between them and the predetermined categories. This layer gives the computer the ability to categorize images on its own.
 - The output layer consists of 10 different nodes, representing the 10 different categories we would sort the images by. Depending on the specific image and the way the computer was trained by the neurons, the output will be the highest predicted category for that particular image.
 
-Optional:
+{{%notice note%}}
```
Comment (ozhang22) [inline content/english/python-tensorflow/building_neural_network.md:58]: Please see if you can fix the spacing for all occurrences here
```suggestion
{{% notice note %}}
```

File: content/english/python-tensorflow/plotting_model.md [hunk 9]
```diff
@@ -7,20 +7,27 @@ weight: 11
...
+Now that we are able to classify the category of an individual piece of clothing using our model, we will visualize these predictions by creating a bar plot for a specified image. 
 
-{{% notice note %}}
+## Predicated Classification 
 
 The `plot_image` function displays an image along with its predicted classification, the confidence level of the model for that category, and the actual classification (which is for reference).
 
-The `plot_value_array` function plots the confidence values generated by the model as a bar graph.
+{{% notice tip %}}
+A confidence level is used to convey how confident or certain we are in our answer being correct. If you have to predict whether it will rain tomorrow, your possible answers are yes or no. What if you're not totally certain in your answer, though? If the weather man predicted rain for tomorrow, you may say you are 90% confident. This means that you are pretty sure it will rain tomorrow, but it also tells us that you think there is a 10% chance you may be wrong.
```
Comment (ozhang22) [inline content/english/python-tensorflow/plotting_model.md:22]:
```suggestion
A **confidence level** is used to convey how confident or certain we are in our answer being correct. If you have to predict whether it will rain tomorrow, your possible answers are yes or no. What if you're not totally certain in your answer, though? If the weather reporter predicted rain for tomorrow, you may say you are 90% confident. This means that you are pretty sure it will rain tomorrow, but it also tells us that you think there is a 10% chance you may be wrong. 
```

File: content/english/python-tensorflow/plotting_model.md [hunk 10]
```diff
@@ -82,8 +107,13 @@ plt.show()
...
 
-Plot the first 25 test images, their predicted labels, and the true labels.
-Color correct predictions in blue and incorrect predictions in red.
+From this example we see that the model predicted the item to be a sandal with 96% confidence. However, we see that the correct identification of the item is a sneaker. The plot shows that the model also predicted this item to be a sneaker with ~ 3% confidence or an ankle boot with ~ 1% confidence. 
+
+## More Examples
+
+For more examples of the model predictions, plot the first 25 test images, their predicted labels, and the true labels. As before, the correct predictions will be shown in blue and the incorrect predictions in red.
+
+Copy this code into Colab:
```
Comment (ozhang22) [inline content/english/python-tensorflow/plotting_model.md:116]:
```suggestion
Copy this code into your Colab notebook: 
```

File: content/english/python-tensorflow/progress_update_2.md [hunk 11]
```diff
@@ -152,6 +134,23 @@ class_names[6] # This is the category name of the image that has the highest pro
...
+If you would like to test the code above, visit <a href="https://colab.research.google.com/drive/1knoCeFRtcUbp1HyKKEQgYw9tgZsUeDk0?usp=sharing" target="_blank">this link</a> 
+
+To edit this code, click on the 'Copy to Drive' button to make a personal copy of this notebook. Make sure you are logged in to your Google account.
+
+{{% notice warning %}}
+### If you are using a Nuevo Google account temporarily
+Once you make a copy, please make sure to replace the "Copy of" with your name, along with the file name. This will be on the top left corner of your notebook.
+{{% /notice%}}
+
+{{% /notice%}}
```
Comment (ozhang22) [inline content/english/python-tensorflow/progress_update_2.md:156]:
```suggestion
{{% /notice %}}

{{% /notice %}}
```

File: content/english/python-tensorflow/wrapup.md [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+You and your colleagues are now able to sort 4 times as many returned clothes! Your supervisor is impressed by this and chooses you as the Employee of the Month!
+
+Thank you for doing this workshop! We hope you enjoyed it and learned the basics of Machine Learning!
+
+{{% notice warning %}}
+#### Are you stuck anywhere in this workshop or want to check your work? 
+Feel free to refer to this: <a href="#" target="_blank">Answer Key</a>
+
+{{% /notice %}}
```
Comment (ozhang22) [inline content/english/python-tensorflow/wrapup.md:21]: We don't usually provide answers to the end of our workshops, but this is up for discussion.

File: content/english/python-tensorflow/activity-2.md [hunk 13]
```diff
@@ -22,26 +24,42 @@ plt.show() #displays the entire plot
...
 ![Shirt](../media/a2progress2.png)
 
 {{% /notice %}}
 
-## Copy the following code onto Google Colab:
+### Normalize Pixel Values:
+
+The program we're going to write takes an input of values between 0 and 1. However, our pixel values are mostly all greater than 1! In fact, the range of values is 0 through 255. How can we change the range of numbers such that it can be inputted into our program? 
+
+We will use a process called "normalization," where we transform these values to make them fit in the range of 0 to 1. More specifically, we will take all of our data and divide it by a singular value so that the range of values can now fit within 0 and 1.
```
Comment (ozhang22) [inline content/english/python-tensorflow/activity-2.md:51]:
```suggestion
We will use a process called **"normalization"**, where we transform these values to make them fit in the range of 0 to 1. More specifically, we will take all of our data and divide it by a singular value so that the range of values can now fit within 0 and 1. 
```

File: content/english/python-tensorflow/activity-2.md [hunk 14]
```diff
@@ -22,26 +24,42 @@ plt.show() #displays the entire plot
...
 
-{{% notice tip %}}
-### Question 2
+#### Question 1:
+After analyzing the shoe, we want to view another item from the clothing sample. Can you figure out how to do that using the previous code block? 
 
-After analyzing the shoe, you now take another clothing item from your sample and do the same thing. Change the index value to whatever you like so you have a different item. 
-Save this value for later as you will use this in Activity 6. 
+{{% notice hint %}}
+Use the code in the previous code block. Then change the index value from of the training images (“train_images[x]”) to get a new image of an item.
```
Comment (ozhang22) [inline content/english/python-tensorflow/activity-2.md:40]:
```suggestion
Use the code in the previous code block. Then change the value of the index in `train_images[x]` to get a new image of an item.   
```

File: content/english/python-tensorflow/activity-1.md [hunk 15]
```diff
@@ -43,8 +47,16 @@ fashion_mnist = keras.datasets.fashion_mnist
 
 (train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data() 
 ```
+
+{{% notice note %}}
+You might notice we are splitting the data into training and testing datasets with their corresponding labels. A training dataset is for our model to learn about the optimal parameters to achieve our tasks, while the test dataset is to validate
+how well our model has learned. This is just like how we learn things: we're being continuously trained and tested to get better!
+{{%/notice%}}
```
Comment (ozhang22) [inline content/english/python-tensorflow/activity-1.md:54]:
```suggestion
{{% notice note %}}
You might notice we are splitting the data into training and testing datasets with their corresponding labels. A **training dataset** is for our model to learn about the optimal parameters to achieve our tasks, while the **test dataset** is to validate how well our model has learned. This is just like how we learn things: we're being continuously trained and tested to get better!
{{% /notice %}}
```

File: content/english/python-tensorflow/activity-2.md [hunk 16]
```diff
@@ -22,26 +24,42 @@ plt.show() #displays the entire plot
 ```
 
 {{% notice tip %}}
-### Question 1
 
-Change the index of the “train_images[x]” and look at the color bar on the right, what is the max pixel value?
+Notice that the image generated has a color bar on the left. This metric is showing that the shoe uses pixel values from 0 to 255. What are pixel values? 
+
+Pixels are the smallest unit of a graphic (image or display) that a screen can represent. Think of your computer screen! There are millions of tiny colored dots that when combined can create a picture of a boat or the text shown on your screen.
```
Comment (ozhang22) [inline content/english/python-tensorflow/activity-2.md:30]:
```suggestion
**Pixels** are the smallest unit of a graphic (image or display) that a screen can represent. Think of your computer screen! There are millions of tiny colored dots that when combined can create a picture of a boat or the text shown on your screen. 
```

File: content/english/python-tensorflow/activity-5.md [hunk 17]
```diff
@@ -9,12 +9,21 @@ draft: false
...
 Now that we have built and trained our model, we want to go ahead and test it against returned clothing items. 
 Let's start off with our first sample item, which was the shoe.
 ```python
+# Initiate a model object with a softmax layer
 probability_model = tf.keras.Sequential([model, tf.keras.layers.Softmax()])
 ```
 
+{{% notice note %}}
+A softmax layer is a type of neural network layer in a maching learning model. The softmax layer turns our inputs, which theoretically could come from a varied range of data structures, be turned into a probability distribution between 0 to 1.
+{{%/notice%}}
```
Comment (ozhang22) [inline content/english/python-tensorflow/activity-5.md:25]:
```suggestion
{{% /notice %}}
```

File: content/english/python-tensorflow/building_neural_network.md [hunk 18]
```diff
@@ -9,27 +9,29 @@ draft: false
 
 <iframe width="100%" height="600px" src="https://www.youtube.com/embed/ttOhB-w8dt0" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
 
-This is the first step in building a neural network, which is actually a tool to allow us to do "machine learning". With this tool, we are training a model/computer to analyze various training images/data, likely labeled in advance, and perform and execute a specific task as a result. In our scenario, we are using pre-labeled images of clothing to train our model to correctly predict the type of clothing for any clothing sample we input.
+This is the first step in building a neural network, which is actually a tool that allows us to do "machine learning". With this tool, we are training a program to analyze images or data, otherwise known as training data.
```
Comment (ozhang22) [inline content/english/python-tensorflow/building_neural_network.md:12]:
```suggestion
This is the first step in building a **neural network**, which is actually a tool that allows us to do **"machine learning"**. With this tool, we are training a program to analyze images or data, otherwise known as **training data**. 
```

File: content/english/python-tensorflow/answer_key.md [hunk 19]
```diff
@@ -217,10 +207,8 @@ array([1.1623413e-08, 2.3745208e-08, 3.5589391e-08, 1.2773025e-09,
 ![](https://i.imgur.com/td1b4Jk.png)
```
Comment (ozhang22) [inline content/english/python-tensorflow/answer_key.md:207]: This is an image, and ideally all images should have alt text that describes the image adequately, for example

```suggestion
![Showing the highest and lowest values of the probability array](https://i.imgur.com/td1b4Jk.png)
```

If you have time, please take a look at adding alt text to all the images in this workshop.

File: content/english/python-tensorflow/_index.md [hunk 20]
```diff
@@ -41,12 +41,6 @@ When running your code for the first time, make sure to run your code by clickin
 
 {{% /notice %}}
 
-## Table of Contents
```
Comment (ozhang22) [inline content/english/python-tensorflow/_index.md:44]: We usually include a table of contents at the beginning of our workshops. Is there a particular reason why this was removed?

**Review by ozhang22 — APPROVED** [review]

File: content/english/python-tensorflow/_index.md (hunk 20 above)
Comment (sofia-geislinger) [reply content/english/python-tensorflow/_index.md:44]: There was nothing listed in the table of contents, I'll go ahead and leave it in though!

Comment (sofia-geislinger) [reply content/english/python-tensorflow/_index.md:44]: wait, whoops it works perfectly fine! My mistake!
